### PR TITLE
ttl: reschedule task to other instances when shrinking worker

### LINKED ISSUE #57990: when shrinking the TTL workers, some expired rows will left undeleted
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. create TTL table with many rows.
2. wait TTL job start
3. set `@@tidb_ttl_scan_worker_count` to a smaller value
4. wait TTL job done
5. check if all expired rows deleted

### 2. What did you expect to see? (Required)

All expired rows should be deleted

### 3. What did you see instead (Required)

some expired rows are not deleted because when shrinking TTL worker, a task will be marked as finished.

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #57990

### What changed and how does it work?

This PR makes some changes:

1. Flush expired rows not deleted in `delWorker` before it stops.
2. When stopping a `scanWorker`, wait for all rows to be deleted before writing its state to task meta.
3. Mark the status of a task back to `waiting` when its worker stops to ensure it can be obtained by other instances again.
4. When task heartbeating, merge the current statistics and the old one that recovered from the task meta.
5. Add more logs.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ttl/ttlworker/task_manager.go [hunk 1]
```diff
@@ -60,7 +60,7 @@ func setTTLTaskFinishedSQL(jobID string, scanID int64, state *cache.TTLTaskState
 	return setTTLTaskFinishedTemplate, []any{now.Format(timeFormat), string(stateStr), jobID, scanID, ownerID}, nil
 }
 
-const updateTTLTaskHeartBeatTempalte = `UPDATE mysql.tidb_ttl_task
+const updateTTLTaskHeartBeatTemplate = `UPDATE mysql.tidb_ttl_task
```
Comment (lcwangchao) [inline pkg/ttl/ttlworker/task_manager.go:63]: typo

File: pkg/ttl/cache/task.go [hunk 2]
```diff
@@ -116,6 +116,9 @@ type TTLTaskState struct {
 	ErrorRows   uint64 `json:"error_rows"`
 
 	ScanTaskErr string `json:"scan_task_err"`
+
+	// When PreviousOwner != "", it means this task is resigned from another owner
+	PreviousOwner string `json:"prev_owner,omitempty"`
```
Comment (YangKeao) [inline pkg/ttl/cache/task.go:121]: How about also print a log in the `lockScanTask` if the `owner` or `previousOwner` is not `NULL`?

It seems that this field is not used by the code logic, so it's only useful for the user to observe the situation, right?

**Review by YangKeao — COMMENTED** [review]
Rest LGTM

**Review by YangKeao — APPROVED** [review]
LGTM

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**